tartufo/scanner.py

```diff
@@ -345,26 +346,26 @@ def entropy_string_is_excluded(self, string: str, line: str, path: str) -> bool:
         )
 
     @lru_cache(maxsize=None)
-    def calculate_entropy(self, data: str, char_set: str) -> float:
+    def calculate_entropy(self, data: str) -> float:
         """Calculate the Shannon entropy for a piece of data.
 
         This essentially calculates the overall probability for each character
-        in `data` to be to be present, based on the characters in `char_set`.
-        By doing this, we can tell how random a string appears to be.
+        in `data` to be to be present. By doing this, we can tell how random a
+        string appears to be.
 
         Borrowed from http://blog.dkbza.org/2007/05/scanning-data-for-entropy-anomalies.html
 
         :param data: The data to be scanned for its entropy
-        :param char_set: The character set used as a basis for the calculation
-        :return: The amount of entropy detected in the data.
+        :return: The amount of entropy detected in the data
         """
         if not data:
             return 0.0
+        frequency = Counter(data)
         entropy = 0.0
-        for char in char_set:
-            prob_x = float(data.count(char)) / len(data)
-            if prob_x > 0:
-                entropy += -prob_x * math.log2(prob_x)
+        float_size = float(len(data))
+        for count in frequency.values():
+            probability = float(count) / float_size
+            entropy += -probability * math.log2(probability)
```

Comment:
So... I'm not saying this is necessarily good or bad, but this is a fundamental change to the way that we are calculating entropy. Instead of calculating how random a string is within a particular character set (base64 or hex), we are calculating it based on how random its characters look compared to themselves.

This idea, or at least very similar ideas, have been previously discussed, but never fully implemented:
* https://github.com/trufflesecurity/truffleHog/issues/168
* https://github.com/Yelp/detect-secrets/issues/234

I think this warrants FAR more investigation and discussion before we dive straight into it with both feet.

One idea that pops to mind is that perhaps we could introduce a new flag along the lines of `--experimental-entropy-detection` (or something less verbose) to let users test it out before we decide this is the right option.

Reply:
Actually, if you look at this closely, the math is bit-for-bit identical. I didn't put in a comment because it's weird enough without referencing code that doesn't even exist any longer, but here's the scoop:

The old algorithm:
1. Loops through each character in `char_set` (i.e. the encoding alphabet)
2. Counts the number of times that character appears in `data` (the target string)
3. Computes `prob_x` -- which makes me think "probability" although it isn't -- as the fraction of the input represented by this character (number of occurrences of character divided by length of target string)
4. Adds this to the running entropy tally using the gonzo equation from the literature

So the first observation is that if `char` doesn't appear in `data`, then `data.count(char)` is 0, which means `prob_x` is 0, which means `char` doesn't contribute to `entropy`.

The second observation is that since only characters in `data` contribute to `entropy`, it doesn't matter what the encoding alphabet is because only the characters actually in `data` count. Notice the only reference to `char_set` is in the `for` statement.

The third observation is that rather than blindly iterating the entire encoding alphabet and scanning the input for each character to count how many times it appears, it would be much better to scan the input once, tallying characters as we go, and then look only at the characters we know are present.

Thus, the new algorithm:
1. Uses `Counter` to quickly (or at least in fewer lines) count up which characters are present and how many times each was used
2. Computes the length of the target string as a float so we don't have to do a type promotion inside every loop pass
3. For each character that is present in `data`, compute the "probability" [sic - really frequency as a fraction of string size] just like before. Note it doesn't even matter what exactly the character is, which is why we iterate on `.values()` -- also note the original algorithm never used `char` (the loop index) except in `data.count()`.
4. Calculates the cumulative `entropy`. Note there is no test on `probability > 0` here because we know it's greater than 0 because we know `count` is greater than 0 because we know it's in `frequency` and we know `Counter` objects only have entries for things that have been counted (at least once).

Looked at differently, both iterate through all of the characters found in `data`...

OLD:
```
for char in [char in char_set if data.count(char) > 0]:
    prob_x = fraction-with-data.count(char)-in-numerator
    add prob_x to entropy
```

NEW:
```
for count in Counter(data).values():  # for char except we don't use char because we only care about count
    prob_x = fraction-with-data.count(char)-in-numerator aka count/size
    add prob_x to entropy
```

Reply:
I was looking for this analogy the entire time I composed the above note, but it didn't come until after I posted it. Regarding the concern that we aren't looking at the representation alphabet any more (not that we were before) and the potential impact on the computation, this basically boils down to "what is the difference between a Big Mac without anchovies and a Big Mac without capers?" The answer, of course, is nothing.

The only reason we care about the alphabet at all -- and not here -- is to know how many bits of data each character represents, so that it is possible to scale the computed entropy value to get an apples-to-apples comparison of the underlying binary bitstream if we had bothered to convert it from base64 or hex representation into raw bits.

To illustrate: How random is the string "11111111"?

If this is a binary number, each digit has a 1-in-2 chance of being "1". If this is a hexadecimal representation, each digit has a 1-in-16 chance of being "1". Either way, it doesn't change the fact that all 8 of the digits are "1".

Reply:
Regarding the independent investigations above, [trufflesecurity/truffleHog#168](https://github.com/trufflesecurity/truffleHog/issues/168) basically does the same derivation, but a bug in the open-coded Counter loop b0rks the results. I posted there with the code here and identified the bug.

Additionally, running the original truffleHog entropy, the proposed (there) single entropy, the old tartufo entropy and the proposed (here) new entropy on the sample strings shows that both originals and the new code proposed here yield exactly identical entropy scores, while the buggy code is, um, buggy.

[Yelp/detect-secrets](https://github.com/Yelp/detect-secrets/issues/234) addresses itself to a different problem, namely the question of what the default entropy detection threshold (aka "sensitivity") should be. This is indeed an interesting question, but is inapropos here to the extent that we maintain exactly the same default threshold as before (and compare the exact same entropy scores to it).